src/middleware/requireChangemakerPermission.ts

```diff
@@ -0,0 +1,38 @@
+import { Response, NextFunction } from 'express';
+import { InputValidationError, UnauthorizedError } from '../errors';
+import { isAuthContext, isId } from '../types';
+import type { Permission } from '../types';
+import type { Request } from 'express';
+
+const requireChangemakerPermission =
+	(permission: Permission) =>
+	(req: Request, res: Response, next: NextFunction) => {
+		if (!isAuthContext(req)) {
+			next(new UnauthorizedError('The request lacks an AuthContext.'));
+			return;
+		}
+		if (req.role?.isAdministrator === true) {
```

Comment:
I'm not sure if this should precede or follow the validation of changemaker ID. It seems right now that check is skipped if you're an admin. However if we check it before checking for admin, we'd be revealing valid and invalid changemaker IDs. Then again, we'd only be revealing those to already authenticated users who would presumably have access to those IDs anyway. Hmm.

Reply:
There is no secret being validated here -- just "is this a GUID" which is part of the API spec.

Reply:
The admin check should come first, for exactly the reason you said! What is the advantage to checking the changemaker ID first?

Reply:
Right. My thought was that the check on changemaker ID here is skipped. But that is OK because it appears to be re-checked later in the handler ~L85 of `putUserChangemakerPermission`. This seems appropriate because the check at L19 below is exclusively for the sake of looking up permissions ~L26.

And as for the secret, right too, it's a format/type check, not checking whether the value exists. Moving along...